### PR TITLE
DAO layer is ready

### PR DESCRIPTION
**Pull request by maksimk99**
1. Add TravisCI
2. Add logging.
3. Unit tests!!!
4. Add DB initialization scripts.

I have one question: should I make unit tests for models?

### CODE REVIEW

**Comment by aleksieiev** [comment]
> should I make unit tests for models?
No. It's required only of model has any business logic more sophisticated than get/set

**Comment by aleksieiev** [comment]
+4,320
Are you really?
It's too much.
Split it to the different commits and PRs please

File: my_checkstyle.xml [hunk 1]
```diff
@@ -0,0 +1,148 @@
+<!DOCTYPE module PUBLIC
```
Comment (aleksieiev) [inline my_checkstyle.xml:1]: 1. I'm pretty confident that it can be separate commit with checkstyle fizes
2. `my_checkstyle.xml`. Why did you change the name? It can be `checkstyle.xml` as well

File: src/main/java/com/spring/rest/config/SpringJDBCConfig.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    /**
+     * Gets named parameter jdbc template.
+     *
+     * @param dataSource is settings connection to database
+     * @return JdbcTemplate template class with a basic set of JDBC operations
+     */
+    @Bean
+    @Autowired
```
Comment (aleksieiev) [inline src/main/java/com/spring/rest/config/SpringJDBCConfig.java:65]: not sure that you need to have it here. As far as I remember it works w/o it

File: src/main/java/com/spring/rest/model/CarCharacteristics.java [hunk 3]
```diff
@@ -0,0 +1,184 @@
+package com.spring.rest.model;
+
+import java.util.Calendar;
+import java.util.Date;
+
+/**
+ * The CarCharacteristics model.
```
Comment (aleksieiev) [inline src/main/java/com/spring/rest/model/CarCharacteristics.java:7]: I want more information here. What kind of abstraction is this class? What is it responsible for?

File: src/main/java/com/spring/rest/dao/CarCharacteristicsDao.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
+package com.spring.rest.dao;
+
+import com.spring.rest.model.CarCharacteristics;
+
+import java.util.List;
+
+/**
+ * The interface provides methods to manage CarCharacteristics model.
+ */
+public interface CarCharacteristicsDao {
```
Comment (aleksieiev) [inline src/main/java/com/spring/rest/dao/CarCharacteristicsDao.java:10]: It's just a philosophical question always.
Why we need to use `getCarsCharacteristics` method? Why it can't be `getAll()` and `get(index)`?
The same for others methods.
No needs to refactor it, just think about it.

File: src/main/java/com/spring/rest/dao/CarModelDao.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
+package com.spring.rest.dao;
+
+import com.spring.rest.model.CarModel;
+
+import java.util.List;
+
+/**
+ * The interface provides methods to manage CarModel model.
```
Comment (aleksieiev) [inline src/main/java/com/spring/rest/dao/CarModelDao.java:8]: `{@see}` section is required

File: src/main/java/com/spring/rest/dao/impl/CarCharacteristicsDaoImpl.java [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+ */
+@Repository
+public class CarCharacteristicsDaoImpl implements CarCharacteristicsDao {
+    /**
+     * SQL query to get car characteristics.
+     */
+    @Value("${car.characteristics.get}")
+    private String GET_CAR_CHARACTERISTICS_SQL;
+    /**
+     * SQL query to get cars characteristics.
```
Comment (aleksieiev) [inline src/main/java/com/spring/rest/dao/impl/CarCharacteristicsDaoImpl.java:32]: `SQL query to get list of cars characteristics.` is better

File: src/main/java/com/spring/rest/dao/impl/CarCharacteristicsDaoImpl.java [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+     */
+    @Value("${cars.characteristics.get}")
+    private String GET_CARS_CHARACTERISTICS_SQL;
```
Comment (aleksieiev) [inline src/main/java/com/spring/rest/dao/impl/CarCharacteristicsDaoImpl.java:35]: GET_LIST_CARS_CHARACTERISTICS_SQL

File: src/main/java/com/spring/rest/dao/impl/CarCharacteristicsDaoImpl.java [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    /**
+     * Delete car characteristics.
+     *
+     * @param index the index
+     */
+    @Override
+    public void deleteCarCharacteristics(final int index) {
+        SqlParameterSource parameters = new MapSqlParameterSource()
+                .addValue("id", index);
+        LOGGER.info("method deleteCarCharacteristics with parameter: {} was called", index);
```
Comment (aleksieiev) [inline src/main/java/com/spring/rest/dao/impl/CarCharacteristicsDaoImpl.java:149]: from my point of view it should be `debug`

File: src/main/java/com/spring/rest/model/mappers/CarCharacteristicsMapper.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+    public CarCharacteristicsMapper(final FuelTypeMapper fuelTypeMapper, final TransmissionMapper transmissionMapper,
+                                    final CarModelMapper carModelMapper) {
+        this.fuelTypeMapper = fuelTypeMapper;
+        this.transmissionMapper = transmissionMapper;
+        this.carModelMapper = carModelMapper;
+    }
+
+
+    @Override
+    public CarCharacteristics mapRow(final ResultSet resultSet, final int i) throws SQLException {
```
Comment (aleksieiev) [inline src/main/java/com/spring/rest/model/mappers/CarCharacteristicsMapper.java:64]: What is `i`?

File: src/main/java/com/spring/rest/model/mappers/CarMakeMapper.java [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+     * The constant CAR_MAKE.
+     */
+    public static final String CAR_MAKE = "car_make";
+    /**
+     * Logger.
+     */
+    private static final Logger LOGGER = LogManager.getLogger(CarMakeMapper.class);
+
+    @Override
+    public CarMake mapRow(final ResultSet resultSet, final int i) throws SQLException {
```
Comment (aleksieiev) [inline src/main/java/com/spring/rest/model/mappers/CarMakeMapper.java:31]: Don't see where is `i` used?

File: src/main/java/com/spring/rest/model/mappers/CarSaleMapper.java [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+        this.userMapper = userMapper;
+        this.carCharacteristicsMapper = carCharacteristicsMapper;
+    }
+
+    @Override
+    public CarSale mapRow(final ResultSet resultSet, final int i) throws SQLException {
+        CarSale carSale = new CarSale(resultSet.getInt(CAR_SALE_ID), resultSet.getBigDecimal(PRICE),
+                resultSet.getDate(DATE), userMapper.mapRow(resultSet, i),
+                carCharacteristicsMapper.mapRow(resultSet, i));
+        LOGGER.info("method mapRow returned: {}", carSale);
```
Comment (aleksieiev) [inline src/main/java/com/spring/rest/model/mappers/CarSaleMapper.java:60]: no needs to log what exactly has been returned - too much logs.
You just can log what row has been mapped

File: src/test/java/com/spring/rest/dao/impl/CarCharacteristicsDaoImplTest.java [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+
+    CarCharacteristicsDao carCharacteristicsDao;
+
+    @Autowired
+    CarCharacteristicsDaoImplTest(CarCharacteristicsDao carCharacteristicsDao) {
+        this.carCharacteristicsDao = carCharacteristicsDao;
+    }
+
+    @Test
+    void getCarsCharacteristics() {
```
Comment (aleksieiev) [inline src/test/java/com/spring/rest/dao/impl/CarCharacteristicsDaoImplTest.java:29]: It's incorrect.
It would be better to remember number of entities before deleting, then remove one of them and check that current number of available entities decreased by 1.

Comment (aleksieiev) [inline src/test/java/com/spring/rest/dao/impl/CarCharacteristicsDaoImplTest.java:29]: Method has name `get`, but it removes something. It should just get number of entities and that's all.
In this case it should be named like `deleteCar...`

File: src/test/java/com/spring/rest/dao/impl/CarCharacteristicsDaoImplTest.java [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+
+    @Test
+    void getCarCharacteristics() {
+        assertEquals(5, carCharacteristicsDao.getCarsCharacteristics().size());
+    }
+
+    @Test
+    void addCarCharacteristics() {
+        int size = carCharacteristicsDao.getCarsCharacteristics().size();
+        int index = carCharacteristicsDao.addCarCharacteristics(new CarCharacteristics(2, new Date(32000),
```
Comment (aleksieiev) [inline src/test/java/com/spring/rest/dao/impl/CarCharacteristicsDaoImplTest.java:48]: Ugly.
Create new class with all these data and send it through get/set.
Then get them from class during check. In this case you avoid typos.

File: src/test/java/com/spring/rest/dao/impl/CarCharacteristicsDaoImplTest.java [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+                133455, new FuelType(2), new Transmission(1), new CarModel(3,
+                new CarMake(2))));
+        assertEquals(size + 1, carCharacteristicsDao.getCarsCharacteristics().size());
+        assertEquals(133455, carCharacteristicsDao.getCarCharacteristics(index).getMileage());
```
Comment (aleksieiev) [inline src/test/java/com/spring/rest/dao/impl/CarCharacteristicsDaoImplTest.java:52]: Get saved object just once and check all fields. In your approach you call DB a few times. What for?

**Review by aleksieiev — APPROVED** [review]
Can be merged, but should be refactored according to the comments. Biggest problem is tests.

File: src/main/java/com/spring/rest/model/mappers/CarCharacteristicsMapper.java (hunk 9 above)
Comment (maksimk99) [reply src/main/java/com/spring/rest/model/mappers/CarCharacteristicsMapper.java:64]: It is a method of the RowMapper interface.
I've just override it.  
the parameter i is the number of the current row

Comment (aleksieiev) [reply src/main/java/com/spring/rest/model/mappers/CarCharacteristicsMapper.java:64]: Why didn't you use it?

Comment (maksimk99) [reply src/main/java/com/spring/rest/model/mappers/CarCharacteristicsMapper.java:64]: It is not necessary for me, because I set a specific value to a specific field
parameter 'i' just shows the current row's number
